doc/man3/SSL_CTX_set_early_cb.pod

```diff
@@ -0,0 +1,92 @@
+
+=head1 SYNOPSIS
+
+ typedef int (*SSL_early_cb_fn) (SSL *s, SSL_EARLY *ectx, int *al, void *arg);
+ void SSL_CTX_set_early_cb(SSL_CTX *c, SSL_early_cb_fn cb, void *arg);
+ int SSL_EARLY_isv2(SSL *s, SSL_EARLY *ectx);
+ unsigned int SSL_EARLY_get0_version(SSL *s, SSL_EARLY *ectx);
+ size_t SSL_EARLY_get0_random(SSL *s, SSL_EARLY *ectx,
+                              const unsigned char **out);
+ size_t SSL_EARLY_get0_session_id(SSL *s, SSL_EARLY *ectx,
+                                  const unsigned char **out);
+ size_t SSL_EARLY_get0_ciphers(SSL *s, SSL_EARLY *ectx,
+                               const unsigned char **out);
+ size_t SSL_EARLY_get0_compression_methods(SSL *s, SSL_EARLY *ectx,
+                                           const unsigned char **out);
+ int SSL_EARLY_get0_ext(SSL *s, SSL_EARLY *ectx, int type,
+                        const unsigned char **out, size_t *outlen);
+
+=head1 DESCRIPTION
+
+SSL_CTX_set_early_cb() sets the callback function, which is automatically
+called during the early stages of ClientHello processing on the server.
+The argument supplied when setting the callback is passed back to the
+callback at runtime.  Error returns from the callback will cause the
+connection to terminate, and callbacks returning failure should indicate
+what alert value is to be sent in the B<al> parameter.
+
+SSL_EARLY_isv2() indicates whether the ClientHello was carried in a
+SSLv2 record and is in the SSLv2 format.
+
+SSL_EARLY_get0_random(), SSL_EARLY_get0_session_id(), SSL_EARLY_get0_ciphers(),
+and SSL_EARLY_get0_compression_methods() provide access to the corresponding
+ClientHello fields, returning the field length and optionally setting an
+out pointer to the octets of that field.
+
+Similarly, SSL_EARLY_get0_ext() provides access to individual extensions
+from the ClientHello on a per-extension basis.  For the provided wire
+protocol extension type value, the extension value and length are returned
+in the output parameters (if present).
+
+=head1 NOTES
+
+The early callback provides a vast window of possibilities for application
+code to affect the TLS handshake.  A primary use of the callback is to
+allow the server to examine the server name indication extension provided
+by the client in order to select an appropriate certificate to present,
+and make other configuration adjustments relevant to that server name
+and its configuration.  Such configuration changes can include swapping out
+the associated SSL_CTX pointer, modifying the server's list of permitted TLS
+versions, changing the server's cipher list, etc.
+
+It is recommended that application callbacks make use of the provided
+SSL_EARLY_* APIs instead of attempting to decode and parse the raw
+ClientHello bytes directly.
```

Comment:
Hmm, this needs to go if we decide to keep the commits that remove this functionality.